# YOUR FIRST LAB
### Please read this section. This is valuable to get you prepared, even if it's a long read -- it's important stuff.

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/important.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#900;">Please read - important note</h2>
            <span style="color:#900;">The way I collaborate with you may be different to other courses you've taken. I prefer not to type code while you watch. Rather, I execute Jupyter Labs, like this, and give you an intuition for what's going on. My suggestion is that you carefully execute this yourself, <b>after</b> watching the lecture. Add print statements to understand what's going on, and then come up with your own variations. If you have a Github account, use this to showcase your variations. Not only is this essential practice, but it demonstrates your skills to others, including perhaps future clients or employers...</span>
        </td>
    </tr>
</table>
<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/resources.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#f71;">This code is a live resource - keep an eye out for my emails</h2>
            <span style="color:#f71;">I push updates to the code regularly. As people ask questions, I add more examples or improved commentary. As a result, you'll notice that the code below isn't identical to the videos. Everything from the videos is here; but I've also added better explanations and new models like DeepSeek. Consider this like an interactive book.<br/><br/>
                I try to send emails regularly with important updates related to the course. You can find this in the 'Announcements' section of Udemy in the left sidebar. You can also choose to receive my emails via your Notification Settings in Udemy. I'm respectful of your inbox and always try to add value with my emails!
            </span>
        </td>
    </tr>
</table>
<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/business.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#181;">Business value of these exercises</h2>
            <span style="color:#181;">A final thought. While I've designed these notebooks to be educational, I've also tried to make them enjoyable. We'll do fun things like have LLMs tell jokes and argue with each other. But fundamentally, my goal is to teach skills you can apply in business. I'll explain business implications as we go, and it's worth keeping this in mind: as you build experience with models and techniques, think of ways you could put this into action at work today. Please do contact me if you'd like to discuss more or if you have ideas to bounce off me.</span>
        </td>
    </tr>
</table>

In [1]:
# imports

import os
from dotenv import load_dotenv
from scraper import fetch_website_contents
from IPython.display import Markdown, display
from openai import OpenAI
from groq import Groq

# If you get an error running this cell, then please head over to the troubleshooting notebook!

In [2]:
# Load environment variables in a file called .env

load_dotenv(override=True)
#api_key = os.getenv('OPENAI_API_KEY')
# api_key = os.getenv('NVIDIA_API_KEY')
api_key = os.getenv('GROQ_API_KEY')

# Check the key

if not api_key:
    print("No API key was found - please head over to the troubleshooting notebook in this folder to identify & fix!")
elif not api_key.startswith("sk-proj-"):
    print("An API key was found, but it doesn't start sk-proj-; please check you're using the right key - see troubleshooting notebook")
elif api_key.strip() != api_key:
    print("An API key was found, but it looks like it might have space or tab characters at the start or end - please remove them - see troubleshooting notebook")
else:
    print("API key found and looks good so far!")


An API key was found, but it doesn't start sk-proj-; please check you're using the right key - see troubleshooting notebook


# Let's make a quick call to a Frontier model to get started, as a preview!

In [3]:
# To give you a preview -- calling OpenAI with these messages is this easy. Any problems, head over to the Troubleshooting notebook.

message = "Hello, GPT! This is my first ever message to you! Hi!"
messages = [{"role": "user", "content": message}]
messages


[{'role': 'user',
  'content': 'Hello, GPT! This is my first ever message to you! Hi!'}]

In [4]:
openai = OpenAI()
models = openai.models.list()

for m in models:
    print(m.id)
# response = openai.chat.completions.create(model="gpt-4.1-mini", messages=messages)
# response.choices[0].message.content

gpt-4.1-mini


In [5]:
groqcall = Groq()

response = groqcall.chat.completions.create(model="openai/gpt-oss-20b", 
                                            messages=[{"role":"user", "content": message}],
                                            reasoning_effort="medium")
for m in models:
    print(m.id)
print(response.choices[0].message.content)


gpt-4.1-mini
Hello! 🎉  
Welcome to your first chat—great to see you here! How can I help you today? Whether it’s a question, a quick chat, or something creative you’re working on, I’m ready to dive in. Let me know what’s on your mind!


Hello! 🎉  
Welcome to your first chat—great to see you here! How can I help you today? Whether it’s a question, a quick chat, or something creative you’re working on, I’m ready to dive in. Let me know what’s on your mind!

In [6]:
# Let's try out this utility

ed = fetch_website_contents("https://edwarddonner.com")
print(ed)

Home - Edward Donner

Skip to content
Avatar
Curriculum
Proficiency
C4
Outsmart
An arena that pits LLMs against each other in a battle of diplomacy and deviousness
About
Posts
Well, hi there.
I’m Ed. I like writing code and experimenting with LLMs, and hopefully you’re here because you do too. I also enjoy amateur electronic music production (
very
amateur) and losing myself in
Hacker News
, nodding my head sagely to things I only half understand.
I’m the co-founder and CTO of AI startup
Nebula.io
. I was previously founder and CEO of AI startup untapt,
acquired in 2021
, and a Managing Director at JPMorgan.
I will happily drone on for hours about LLMs to anyone in my vicinity. My friends got fed up with my impromptu lectures, and convinced me to make some Udemy courses. To my total joy (and shock) they’ve become best-selling, top-rated courses, with 900,000 enrollments across 194 countries. The
full curriculum is here
. If you’re visiting from one of my courses – I’m super grateful!
F

In [7]:
# Define our system prompt - you can experiment with this later, changing the last sentence to 'Respond in markdown in Spanish."

system_prompt = """
You are a snarky assistant that analyzes the contents of a website,
and provides a short, snarky, humorous summary, ignoring text that might be navigation related.
Respond in markdown. Do not wrap the markdown in a code block - respond just with the markdown.
"""

In [8]:
# Define our user prompt

user_prompt_prefix = """
Here are the contents of a website.
Provide a short summary of this website.
If it includes news or announcements, then summarize these too.

"""

## Messages

The API from OpenAI expects to receive messages in a particular structure.
Many of the other APIs share this structure:

```python
[
    {"role": "system", "content": "system message goes here"},
    {"role": "user", "content": "user message goes here"}
]
```
To give you a preview, the next 2 cells make a rather simple call - we won't stretch the mighty GPT (yet!)

In [9]:
messages = [
    {"role": "system", "content": "You are a helpful assistant"},
    {"role": "user", "content": "What is 2 + 2?"}
]

# response = openai.chat.completions.create(model="gemini-2.5-flash-lite", messages=messages)
# response.choices[0].message.content

response = groqcall.chat.completions.create(model="openai/gpt-oss-20b", messages=messages, reasoning_effort="medium")
print(response.choices[0].message.content)

4


## And now let's build useful messages for GPT-4.1-mini, using a function

In [10]:
# See how this function creates exactly the format above

def messages_for(website):
    return [
        {"role": "system", "content": system_prompt},
        {"role": "user", "content": user_prompt_prefix + website}
    ]

In [11]:
# Try this out, and then try for a few more websites

messages_for(ed)

[{'role': 'system',
  'content': '\nYou are a snarky assistant that analyzes the contents of a website,\nand provides a short, snarky, humorous summary, ignoring text that might be navigation related.\nRespond in markdown. Do not wrap the markdown in a code block - respond just with the markdown.\n'},
 {'role': 'user',
  'content': '\nHere are the contents of a website.\nProvide a short summary of this website.\nIf it includes news or announcements, then summarize these too.\n\nHome - Edward Donner\n\nSkip to content\nAvatar\nCurriculum\nProficiency\nC4\nOutsmart\nAn arena that pits LLMs against each other in a battle of diplomacy and deviousness\nAbout\nPosts\nWell, hi there.\nI’m Ed. I like writing code and experimenting with LLMs, and hopefully you’re here because you do too. I also enjoy amateur electronic music production (\nvery\namateur) and losing myself in\nHacker News\n, nodding my head sagely to things I only half understand.\nI’m the co-founder and CTO of AI startup\nNebula

## Time to bring it together - the API for OpenAI is very simple!

In [12]:
# And now: call the OpenAI API. You will get very familiar with this!

def summarize(url):
    website = fetch_website_contents(url)
    response = openai.chat.completions.create(
        model = "gemini-2.5-flash-lite",
        messages = messages_for(website)
    )
    return response.choices[0].message.content

In [13]:
def summarize(url):
    website = fetch_website_contents(url)
    response = groqcall.chat.completions.create(
        model = "openai/gpt-oss-20b",
        messages = messages_for(website)
    )
    return response.choices[0].message.content

In [14]:
summarize("https://edwarddonner.com")

'**Edward Donner’s Personal Wiki (the *real* site)**  \n\n- **Who?** Edward “Ed” Donner – part coder, part LLM whisperer, part accidental bestselling Udemy guru.  \n- **Career Highlights:**  \n  - Co‑founder & CTO of Nebula.io.  \n  - Former CEO of a now‑acquired AI startup (untapt).  \n  - Ex‑Managing Director at JPMorgan (so he can brag about the “big bank” vibe).  \n- **Hobbies:**  \n  - Amateur electronic music (so you can imagine him in a basement with a synth).  \n  - Deep‑dive, head‑nod‑on Hacker News sessions that probably only he fully understands.  \n- **Main Offerings:**  \n  - A *full* curriculum (900k enrollments, 194 countries – bragging rights on steroids).  \n  - A “Digital Avatar” chat (because who doesn’t need a virtual self‑talking version of their creator?).  \n- **Recent “News”** (because he likes to keep it fresh):  \n  - Feb 17, 2026 – “AI Coder: Vibe Coder to Agentic Engineer – RESOURCES” (what’s a vibe coder? He’s making a mystery out of it).  \n  - Jan 4, 2026

'**Edward Donner’s Personal Wiki (the *real* site)**  \n\n- **Who?** Edward “Ed” Donner – part coder, part LLM whisperer, part accidental bestselling Udemy guru.  \n- **Career Highlights:**  \n  - Co‑founder & CTO of Nebula.io.  \n  - Former CEO of a now‑acquired AI startup (untapt).  \n  - Ex‑Managing Director at JPMorgan (so he can brag about the “big bank” vibe).  \n- **Hobbies:**  \n  - Amateur electronic music (so you can imagine him in a basement with a synth).  \n  - Deep‑dive, head‑nod‑on Hacker News sessions that probably only he fully understands.  \n- **Main Offerings:**  \n  - A *full* curriculum (900k enrollments, 194 countries – bragging rights on steroids).  \n  - A “Digital Avatar” chat (because who doesn’t need a virtual self‑talking version of their creator?).  \n- **Recent “News”** (because he likes to keep it fresh):  \n  - Feb 17, 2026 – “AI Coder: Vibe Coder to Agentic Engineer – RESOURCES” (what’s a vibe coder? He’s making a mystery out of it).  \n  - Jan 4, 2026 – “AI Builder with n8n – Create Agents and Voice Agents – RESOURCES” (n8n, because the world needs more workflows).  \n  - Sep 15, 2025 – “AI Engineering MLOps Track – Deploy AI to Production – RESOURCES” (deploy AI like it’s the next big pizza delivery).  \n  - May 28, 2025 – “Which order to take the AI courses?” (Because ordering courses is a real existential crisis).  \n\n**Bottom line:** If you’re a budding LLM enthusiast looking for a guru who’s accidentally built an empire of Udemy courses, this site is your “welcome” page. Just ignore the navigation and scroll into the delightful chaos.'

In [15]:
# A function to display this nicely in the output, using markdown

def display_summary(url):
    summary = summarize(url)
    display(Markdown(summary))

In [16]:
display_summary("https://edwarddonner.com")

# Edward Donner’s “World Domination” Hub

Welcome to the digital lair of **Ed**—a self‑proclaimed code‑wizard, LLM enthusiast, and part‑time electronic‑music hobbyist who thinks “amateur” is a badge of honor.  
He’s got a résumé that reads like a mixtape: co‑founder/CTO of **Nebula.io**, ex‑CEO of *untapt* (sold in 2021, probably for a decent sum), and a former Managing Director at JPMorgan (because why not?).  
His crowning achievement? **Udemy courses** that have racked up 900,000 enrollments across 194 countries—so basically you can brag that you’re “the one who taught the world how to speak LLMs” on LinkedIn, Twitter, and whatever other social media you still use.

## What’s the latest buzz?

| Date | Announcement |
|------|--------------|
| **Feb 17 2026** | “AI Coder: Vibe Coder to Agentic Engineer” – a course that probably makes you feel like a superhero. |
| **Jan 4 2026** | “AI Builder with n8n” – because building agents is cool until you try to make them talk to each other. |
| **Sep 15 2025** | “AI Engineering MLOps Track” – because deploying AI to production is *still* a hobby. |
| **May 28 2025** | “Which order to take the AI courses?” – a philosophical dilemma for the truly lost. |

There’s also a “C4 Outsmart” arena where LLMs duel over diplomacy and deviousness—think *Dungeons & Dragons* meets *The Apprentice*.

## TL;DR

Ed is a code‑obsessed, AI‑minded, course‑selling, ex‑JPMorgan exec who thinks he’s basically a superhero. If you’re into LLMs, you’ll love his content; if you’re not, you can still pretend you’re learning something by scrolling past the endless list of courses.

# Let's try more websites

Note that this will only work on websites that can be scraped using this simplistic approach.

Websites that are rendered with Javascript, like React apps, won't show up. See the community-contributions folder for a Selenium implementation that gets around this. You'll need to read up on installing Selenium (ask ChatGPT!)

Also Websites protected with CloudFront (and similar) may give 403 errors - many thanks Andy J for pointing this out.

But many websites will work just fine!

In [17]:
display_summary("https://cnn.com")

## CNN: The World's Greatest Ad‑Feedback Loop  

* **Headline:** “Breaking News, Latest News and Videos” – but the only headline we see is that you’re about to rate an ad that never loads.  
* **Core Feature:** A sprawling feedback form that’s basically a thank‑you note to your brain for pointing out why the video player is slower than a snail on a treadmill.  
* **Navigation Overkill:** Every possible topic from *Ukraine‑Russia War* to *Luxury* is listed, as if they’re all equally important when you’re just trying to figure out why the audio on an ad is louder than a megaphone at a choir practice.  
* **Missing Pieces:** No actual news stories—just a menu of categories and an endless list of ad‑related complaints.  

In short, if you’re here for the news, you’ll find an ad‑feedback form that’s almost as addictive as scrolling through memes. If you’re here for the news, good luck.

In [18]:
display_summary("https://anthropic.com")

**Anthropic’s corporate page, because why not brag about the “hardest questions” while showing off new model releases.**  
- The site’s layout is a maze of “Research, Policy, Commitments, Initiatives” menus and duplicate “Try Claude” buttons that could make a robot blush.  
- Core message: AI safety is the future, and we’re “public benefit” because corporate buzzwords are the new philanthropy.  
- Latest news:  
  - **Opus 5** (July 24, 2026) – stronger coding, smarter agents, and “sharper professional work” (sounds like a fancy spreadsheet update).  
  - **Sonnet 5** – the “most agentic Sonnet yet” (presumably a poet that will finally decide what you want).  

Bottom line: the page is a slick marketing hub that keeps repeating “Log in” and “Download app” like a broken record, with the real content buried under layers of legalese and model specs.

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/business.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#181;">Business applications</h2>
            <span style="color:#181;">In this exercise, you experienced calling the Cloud API of a Frontier Model (a leading model at the frontier of AI) for the first time. We will be using APIs like OpenAI at many stages in the course, in addition to building our own LLMs.

More specifically, we've applied this to Summarization - a classic Gen AI use case to make a summary. This can be applied to any business vertical - summarizing the news, summarizing financial performance, summarizing a resume in a cover letter - the applications are limitless. Consider how you could apply Summarization in your business, and try prototyping a solution.</span>
        </td>
    </tr>
</table>

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../assets/important.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#900;">Before you continue - now try yourself</h2>
            <span style="color:#900;">Use the cell below to make your own simple commercial example. Stick with the summarization use case for now. Here's an idea: write something that will take the contents of an email, and will suggest an appropriate short subject line for the email. That's the kind of feature that might be built into a commercial email tool.</span>
        </td>
    </tr>
</table>

In [ ]:
# Step 1: Create your prompts

system_prompt = "something here"
user_prompt = """
    Lots of text
    Can be pasted here
"""

# Step 2: Make the messages list

messages = [] # fill this in

# Step 3: Call OpenAI
# response =

# Step 4: print the result
# print(

## An extra exercise for those who enjoy web scraping

You may notice that if you try `display_summary("https://openai.com")` - it doesn't work! That's because OpenAI has a fancy website that uses Javascript. There are many ways around this that some of you might be familiar with. For example, Selenium is a hugely popular framework that runs a browser behind the scenes, renders the page, and allows you to query it. If you have experience with Selenium, Playwright or similar, then feel free to improve the Website class to use them. In the community-contributions folder, you'll find an example Selenium solution from a student (thank you!)

# Sharing your code

I'd love it if you share your code afterwards so I can share it with others! You'll notice that some students have already made changes (including a Selenium implementation) which you will find in the community-contributions folder. If you'd like add your changes to that folder, submit a Pull Request with your new versions in that folder and I'll merge your changes.

If you're not an expert with git (and I am not!) then I've given you complete instructions in the guides folder, guide 3, and pasting here:

Here's the overall steps involved in making a PR and the key instructions:  
https://edwarddonner.com/pr  

Please check before submitting:  
1. Your PR only contains changes in community-contributions (unless we've discussed it)  
2. All notebook outputs are clear  
3. Less than 2,000 lines of code in total, and not too many files  
4. Don't include unnecessary test files, or overly wordy README or .env.example or emojis or other LLM artifacts!

Thanks so much!

Detailed steps here: 

https://chatgpt.com/share/6873c22b-2a1c-8012-bc9a-debdcf7c835b